"""
gRPC Client: Request Service

–ü—Ä–µ–¥–º–µ—Ç–Ω–∞—è –æ–±–ª–∞—Å—Ç—å: –ü–°–û ¬´–Æ–≥–æ-–ó–∞–ø–∞–¥¬ª
"""
import grpc
import sys
import os

# –î–æ–±–∞–≤–ª–µ–Ω–∏–µ generated/ –≤ sys.path
sys.path.append(os.path.join(os.path.dirname(__file__), '..'))

from generated import request_service_pb2, request_service_pb2_grpc


def create_request(stub):
    """
    Unary RPC: –°–æ–∑–¥–∞—Ç—å –∑–∞—è–≤–∫—É
    
    Args:
        stub: RequestServiceStub
    
    Returns:
        request_id: str
    """
    print("\n=== CreateRequest ===")
    
    zone = request_service_pb2.Zone(
        name="North",
        lat_min=52.0,
        lat_max=52.5,
        lon_min=23.5,
        lon_max=24.0
    )
    
    request = request_service_pb2.CreateRequestRequest(
        coordinator_id="COORD-1",
        zone=zone
    )
    
    response = stub.CreateRequest(request)
    
    if response.status == "SUCCESS":
        print(f"‚úÖ Request created: {response.request_id}")
        return response.request_id
    else:
        print(f"‚ùå Error: {response.error_message}")
        return None


def get_request(stub, request_id):
    """
    Unary RPC: –ü–æ–ª—É—á–∏—Ç—å –∑–∞—è–≤–∫—É –ø–æ ID
    
    Args:
        stub: RequestServiceStub
        request_id: str
    """
    print(f"\n=== GetRequest({request_id}) ===")
    
    request = request_service_pb2.GetRequestRequest(request_id=request_id)
    
    try:
        response = stub.GetRequest(request)
        
        if response.found:
            req = response.request
            print(f"üì¶ Request ID: {req.request_id}")
            print(f"   Coordinator: {req.coordinator_id}")
            print(f"   Zone: {req.zone.name} ({req.zone.lat_min}, {req.zone.lat_max})")
            print(f"   Status: {req.status}")
            print(f"   Created: {req.created_at}")
        else:
            print("‚ùå Request not found")
    
    except grpc.RpcError as e:
        print(f"‚ùå RPC Error: {e.code()} - {e.details()}")


def list_requests(stub, status_filter=None):
    """
    Unary RPC: –°–ø–∏—Å–æ–∫ –∑–∞—è–≤–æ–∫
    
    Args:
        stub: RequestServiceStub
        status_filter: str ("ACTIVE", "COMPLETED", None)
    """
    print(f"\n=== ListRequests(filter={status_filter or 'ALL'}) ===")
    
    request = request_service_pb2.ListRequestsRequest(
        status_filter=status_filter or "",
        limit=10
    )
    
    response = stub.ListRequests(request)
    
    print(f"üìã Found {response.total_count} requests:")
    for req in response.requests:
        print(f"  ‚Üí {req.request_id} ({req.status}) - {req.zone.name}")


def activate_request(stub, request_id):
    """
    Unary RPC: –ê–∫—Ç–∏–≤–∏—Ä–æ–≤–∞—Ç—å –∑–∞—è–≤–∫—É
    
    Args:
        stub: RequestServiceStub
        request_id: str
    """
    print(f"\n=== ActivateRequest({request_id}) ===")
    
    request = request_service_pb2.ActivateRequestRequest(request_id=request_id)
    
    response = stub.ActivateRequest(request)
    
    if response.success:
        print(f"üöÄ Request activated: {request_id}")
    else:
        print(f"‚ùå Error: {response.error_message}")


def stream_active_requests(stub):
    """
    Server-side Streaming: –°—Ç—Ä–∏–º –∞–∫—Ç–∏–≤–Ω—ã—Ö –∑–∞—è–≤–æ–∫
    
    Args:
        stub: RequestServiceStub
    """
    print("\n=== StreamActiveRequests ===")
    
    request = request_service_pb2.StreamActiveRequestsRequest()
    
    print("üì° Streaming active requests (press Ctrl+C to stop)...")
    
    try:
        for req in stub.StreamActiveRequests(request):
            print(f"  ‚Üí {req.request_id} ({req.status}) - Zone: {req.zone.name}")
    
    except KeyboardInterrupt:
        print("\n‚èπÔ∏è  Stopped streaming")
    
    except grpc.RpcError as e:
        print(f"‚ùå RPC Error: {e.code()} - {e.details()}")


def run():
    """–û—Å–Ω–æ–≤–Ω–∞—è –ª–æ–≥–∏–∫–∞ –∫–ª–∏–µ–Ω—Ç–∞"""
    # –ü–æ–¥–∫–ª—é—á–µ–Ω–∏–µ –∫ —Å–µ—Ä–≤–µ—Ä—É
    with grpc.insecure_channel('localhost:50051') as channel:
        stub = request_service_pb2_grpc.RequestServiceStub(channel)
        
        # 1. –°–æ–∑–¥–∞–Ω–∏–µ –∑–∞—è–≤–∫–∏
        request_id = create_request(stub)
        
        if request_id:
            # 2. –ü–æ–ª—É—á–µ–Ω–∏–µ –∑–∞—è–≤–∫–∏
            get_request(stub, request_id)
            
            # 3. –ê–∫—Ç–∏–≤–∞—Ü–∏—è –∑–∞—è–≤–∫–∏
            activate_request(stub, request_id)
        
        # 4. –°–ø–∏—Å–æ–∫ –≤—Å–µ—Ö –∑–∞—è–≤–æ–∫
        list_requests(stub)
        
        # 5. –°–ø–∏—Å–æ–∫ —Ç–æ–ª—å–∫–æ –∞–∫—Ç–∏–≤–Ω—ã—Ö –∑–∞—è–≤–æ–∫
        list_requests(stub, status_filter="ACTIVE")
        
        # 6. –°—Ç—Ä–∏–º –∞–∫—Ç–∏–≤–Ω—ã—Ö –∑–∞—è–≤–æ–∫ (—Ä–∞—Å–∫–æ–º–º–µ–Ω—Ç–∏—Ä–æ–≤–∞—Ç—å –¥–ª—è —Ç–µ—Å—Ç–∏—Ä–æ–≤–∞–Ω–∏—è)
        # stream_active_requests(stub)


if __name__ == '__main__':
    run()
